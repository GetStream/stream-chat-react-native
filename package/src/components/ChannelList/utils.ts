import type { Channel, StreamChat } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../types/types';

type MoveParameters<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = {
  channels: Channel<StreamChatGenerics>[];
  cid: string;
};

export const moveChannelUp = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  channels = [],
  cid,
}: MoveParameters<StreamChatGenerics>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = {
  client: StreamChat<StreamChatGenerics>;
  id: string;
  type: string;
};

export const getChannel = async <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  client,
  id,
  type,
}: GetParameters<StreamChatGenerics>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
