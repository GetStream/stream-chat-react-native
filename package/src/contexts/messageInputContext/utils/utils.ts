import { lookup } from 'mime-types';
import type { FileUploadConfig } from 'stream-chat';

import { Asset, File } from '../../../types/types';

export const MAX_FILE_SIZE_TO_UPLOAD = 100 * 1024 * 1024; // 100 MB

type CheckUploadPermissionsParams = {
  config: FileUploadConfig;
  file: File | Partial<Asset>;
};

/**
 * This utility function checks if the file upload is allowed based on the file upload config.
 * @param Object File upload config and file to check
 * @returns
 */
export const isUploadAllowed = ({ config, file }: CheckUploadPermissionsParams) => {
  const {
    allowed_file_extensions,
    allowed_mime_types,
    blocked_file_extensions,
    blocked_mime_types,
  } = config;

  if (allowed_file_extensions?.length) {
    const allowed = allowed_file_extensions.some((fileExtension: string) =>
      file.name?.toLowerCase().endsWith(fileExtension.toLowerCase()),
    );

    if (!allowed) {
      return false;
    }
  }

  if (blocked_file_extensions?.length) {
    const blocked = blocked_file_extensions.some((fileExtension: string) =>
      file.name?.toLowerCase().endsWith(fileExtension.toLowerCase()),
    );

    if (blocked) {
      return false;
    }
  }

  if (allowed_mime_types?.length) {
    if (file.name) {
      const fileMimeType = lookup(file.name) as string;
      const allowed = allowed_mime_types.some(
        (mimeType: string) => fileMimeType.toLowerCase() === mimeType.toLowerCase(),
      );

      if (!allowed) {
        return false;
      }
    }
  }

  if (blocked_mime_types?.length) {
    if (file.name) {
      const fileMimeType = lookup(file.name) as string;
      const blocked = blocked_mime_types.some(
        (mimeType: string) => fileMimeType.toLowerCase() === mimeType.toLowerCase(),
      );

      if (blocked) {
        return false;
      }
    }
  }

  return true;
};

/**
 * This utility function prettifies the file size.
 * @param bytes The bytes of the file
 * @param precision The precision to which the file size should be rounded
 * @returns
 */
export function prettifyFileSize(bytes: number, precision = 3) {
  const units = ['B', 'kB', 'MB', 'GB'];
  const exponent = Math.min(Math.floor(Math.log(bytes) / Math.log(1024)), units.length - 1);
  const mantissa = bytes / 1024 ** exponent;
  return `${mantissa.toPrecision(precision)} ${units[exponent]}`;
}
