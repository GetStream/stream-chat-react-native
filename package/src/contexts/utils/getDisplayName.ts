import type { ComponentType } from 'react';

import type { UnknownType } from '../../types/types';

export const getDisplayName = <P extends UnknownType>(Component: ComponentType<P>) =>
  Component.displayName || Component.name || 'Component';
