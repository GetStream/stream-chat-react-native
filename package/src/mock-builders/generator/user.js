import { v4 as uuidv4 } from 'uuid';

export const getUserDefaults = () => ({
  banned: false,
  created_at: '2020-04-27T13:39:49.331742Z',
  id: uuidv4(),
  image: uuidv4(),
  name: uuidv4(),
  online: false,
  role: 'user',
  updated_at: '2020-04-27T13:39:49.332087Z',
});

export const generateUser = (options = {}) => ({
  ...getUserDefaults(),
  ...options,
});

const staticUsers = [
  // By the order of...
  generateUser({
    id: 'tommy',
    image: 'https://i.imgur.com/ytIyojl.png',
    name: 'Thomas',
  }),
  generateUser({
    id: 'arthur',
    image: 'https://i.imgur.com/LuuGvh0.png',
    name: 'Arthur',
  }),
  generateUser({
    id: 'john',
    image: 'https://i.imgur.com/LbtxRjf.png',
    name: 'John',
  }),
  generateUser({
    id: 'finn',
    image: 'https://i.imgur.com/spueyAP.png',
    name: 'Finn',
  }),
];

export const generateStaticUser = (userNumber) => {
  if (userNumber - 1 > staticUsers.length) {
    throw new Error(`Tried getting a static user that doesn't exist.
Index: ${userNumber} , number of users: ${staticUsers.length}`);
  }
  return staticUsers[userNumber];
};
