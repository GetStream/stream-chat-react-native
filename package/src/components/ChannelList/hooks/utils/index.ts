import { Channel, ChannelSortBase } from 'stream-chat';

import { DefaultStreamChatGenerics } from '../../../../types/types';
import { ChannelListProps } from '../../ChannelList';

export const isChannelPinned = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  channel: Channel<StreamChatGenerics>,
) => {
  if (!channel) {
    return false;
  }

  const member = channel.state.membership;

  return !!member?.pinned_at;
};

export const isChannelArchived = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  channel: Channel<StreamChatGenerics>,
) => {
  if (!channel) {
    return false;
  }

  const member = channel.state.membership;

  return !!member?.archived_at;
};

export const shouldConsiderArchivedChannels = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  filters: ChannelListProps<StreamChatGenerics>['filters'],
) => {
  if (!filters) {
    return false;
  }

  return typeof filters.archived === 'boolean';
};

export const extractSortValue = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  atIndex,
  sort,
  targetKey,
}: {
  atIndex: number;
  targetKey: keyof ChannelSortBase<StreamChatGenerics>;
  sort?: ChannelListProps<StreamChatGenerics>['sort'];
}) => {
  if (!sort) {
    return null;
  }
  let option: null | ChannelSortBase<StreamChatGenerics> = null;

  if (Array.isArray(sort)) {
    option = sort[atIndex] ?? null;
  } else {
    let index = 0;
    for (const key in sort) {
      if (index !== atIndex) {
        index++;
        continue;
      }

      if (key !== targetKey) {
        return null;
      }

      option = sort;

      break;
    }
  }

  return option?.[targetKey] ?? null;
};

/**
 * Returns true only if `{ pinned_at: -1 }` or `{ pinned_at: 1 }` option is first within the `sort` array.
 */
export const shouldConsiderPinnedChannels = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  sort: ChannelListProps<StreamChatGenerics>['sort'],
) => {
  const value = extractSortValue({
    atIndex: 0,
    sort,
    targetKey: 'pinned_at',
  });

  if (typeof value !== 'number') {
    return false;
  }

  return Math.abs(value) === 1;
};

export function findPinnedAtSortOrder<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({ sort }: { sort: ChannelListProps<StreamChatGenerics>['sort'] }) {
  if (!sort) {
    return null;
  }

  if (Array.isArray(sort)) {
    const [option] = sort;

    if (!option?.pinned_at) {
      return null;
    }

    return option.pinned_at;
  } else {
    if (!sort.pinned_at) {
      return null;
    }

    return sort.pinned_at;
  }
}

export function findLastPinnedChannelIndex<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({ channels }: { channels: Channel<StreamChatGenerics>[] }) {
  let lastPinnedChannelIndex: number | null = null;

  for (const channel of channels) {
    if (!isChannelPinned(channel)) {
      break;
    }

    if (typeof lastPinnedChannelIndex === 'number') {
      lastPinnedChannelIndex++;
    } else {
      lastPinnedChannelIndex = 0;
    }
  }

  return lastPinnedChannelIndex;
}
