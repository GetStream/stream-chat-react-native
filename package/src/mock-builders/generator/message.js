import { v4 as uuidv4, v5 as uuidv5 } from 'uuid';

import { generateUser } from './user';

export const generateMessage = (options = {}) => {
  const timestamp =
    options.timestamp || new Date(new Date().getTime() - Math.floor(Math.random() * 100000));

  return {
    attachments: [],
    created_at: timestamp,
    html: '<p>regular</p>',
    id: uuidv4(),
    text: uuidv4(),
    type: 'regular',
    updated_at: timestamp.toString(),
    user: generateUser(),
    ...options,
  };
};

const StreamReactNativeNamespace = '9b244ee4-7d69-4d7b-ae23-cf89e9f7b035';
export const generateStaticMessage = (seed, options, date) =>
  generateMessage({
    created_at: date || '2020-04-27T13:39:49.331742Z',
    id: uuidv5(seed, StreamReactNativeNamespace),
    text: seed,
    updated_at: date || '2020-04-27T13:39:49.331742Z',
    ...options,
  });
