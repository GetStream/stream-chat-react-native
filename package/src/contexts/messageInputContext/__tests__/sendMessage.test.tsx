import React, { PropsWithChildren } from 'react';
import { act } from 'react-test-renderer';

import { renderHook } from '@testing-library/react-hooks';

import type { MessageType } from '../../../components/MessageList/hooks/useMessageList';
import {
  generateFileUploadPreview,
  generateImageUploadPreview,
} from '../../../mock-builders/generator/attachment';
import { generateMessage } from '../../../mock-builders/generator/message';
import { generateUser } from '../../../mock-builders/generator/user';
import type { DefaultStreamChatGenerics } from '../../../types/types';
import { FileState } from '../../../utils/utils';
import {
  InputMessageInputContextValue,
  MessageInputContextValue,
  MessageInputProvider,
  useMessageInputContext,
} from '../MessageInputContext';

type WrapperType<StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics> =
  Partial<InputMessageInputContextValue<StreamChatGenerics>>;

const Wrapper = <StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics>({
  children,
  ...rest
}: PropsWithChildren<WrapperType<StreamChatGenerics>>) => (
  <MessageInputProvider
    value={
      {
        ...rest,
      } as MessageInputContextValue<StreamChatGenerics>
    }
  >
    {children}
  </MessageInputProvider>
);

describe("MessageInputContext's sendMessage", () => {
  const message: boolean | MessageType<DefaultStreamChatGenerics> = generateMessage({
    created_at: 'Sat Jul 02 2022 23:55:13 GMT+0530 (India Standard Time)',
    id: '7a85f744-cc89-4f82-a1d4-5456432cc8bf',
    updated_at: 'Sat Jul 02 2022 23:55:13 GMT+0530 (India Standard Time)',
    user: generateUser({
      id: '5d6f6322-567e-4e1e-af90-97ef1ed5cc23',
      image: 'fc86ddcb-bac4-400c-9afd-b0c0a1c0cd33',
      name: '50cbdd0e-ca7e-4478-9e2c-be0f1ac6a995',
    }),
  }) as unknown as MessageType<DefaultStreamChatGenerics>;

  it('exit sendMessage when file upload status failed', async () => {
    const files = generateFileUploadPreview({ state: FileState.UPLOAD_FAILED });
    const images = generateImageUploadPreview({ state: FileState.UPLOAD_FAILED });
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        editing: true,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setFileUploads([files]);
      result.current.setImageUploads([images]);
    });

    act(() => {
      result.current.sendMessage();
    });

    await expect(result.current.sending.current).toBeFalsy();
  });

  it('exit sendMessage when image upload status is uploading', async () => {
    const images = generateImageUploadPreview({ state: FileState.UPLOADING });

    const { rerender, result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        editing: true,
        sendImageAsync: true,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setImageUploads([images]);
      result.current.setMentionedUsers([]);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    rerender({ editing: false, sendImageAsync: true });

    await expect(result.current.asyncIds).toHaveLength(1);
    await expect(result.current.sending.current).toBeFalsy();
  });

  it('exit sendMessage when image upload status is uploading and sendImageAsync is available', async () => {
    const images = generateImageUploadPreview({ state: FileState.UPLOADING });

    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        editing: true,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setImageUploads([images]);
      result.current.setMentionedUsers([]);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    await expect(result.current.sending.current).toBeFalsy();
  });

  it('exit sendMessage when file upload status is uploading', async () => {
    const files = generateFileUploadPreview({ state: FileState.UPLOADING });
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        editing: true,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setFileUploads([files]);
      result.current.setMentionedUsers([]);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    await expect(result.current.sending.current).toBeFalsy();
  });

  it('exit sendMessage when image upload status is uploaded successfully', () => {
    const sendMessageMock = jest.fn();
    const clearQuotedMessageStateMock = jest.fn();
    const images = [
      generateImageUploadPreview({ state: FileState.UPLOADED }),
      generateImageUploadPreview({ state: FileState.FINISHED }),
    ];
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        clearQuotedMessageState: clearQuotedMessageStateMock,
        editing: true,
        quotedMessage: false,
        sendMessage: sendMessageMock,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setImageUploads(images);
      result.current.setMentionedUsers(['dummy1', 'dummy2']);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    expect(sendMessageMock.mock.calls[0][0]).toMatchSnapshot();
    expect(clearQuotedMessageStateMock).toHaveBeenCalled();
    expect(result.current.sending.current).toBeFalsy();
    expect(result.current.fileUploads.length).toBe(0);
    expect(result.current.imageUploads.length).toBe(0);
    expect(result.current.mentionedUsers.length).toBe(0);
  });

  it('exit sendMessage when image upload has an error and catch block is executed', () => {
    const setQuotedMessageStateMock = jest.fn();
    const clearQuotedMessageStateMock = jest.fn();
    const generatedQuotedMessage: boolean | MessageType<DefaultStreamChatGenerics> = message;
    const images = [
      generateImageUploadPreview({ state: FileState.UPLOADED }),
      generateImageUploadPreview({ state: FileState.FINISHED }),
    ];
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        clearQuotedMessageState: clearQuotedMessageStateMock,
        editing: true,
        quotedMessage: generatedQuotedMessage,
        setQuotedMessageState: setQuotedMessageStateMock,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setImageUploads(images);
      result.current.setMentionedUsers(['dummy1', 'dummy2']);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    expect(setQuotedMessageStateMock).toHaveBeenCalled();
    expect(result.current.sending.current).toBeFalsy();
  });

  it('exit sendMessage when edit message is not boolean image upload status is uploaded successfully', () => {
    const sendMessageMock = jest.fn();
    const clearQuotedMessageStateMock = jest.fn();
    const clearEditingStateMock = jest.fn();
    const editMessageMock = jest.fn().mockResolvedValue({ data: {} });
    const images = generateImageUploadPreview({ state: FileState.UPLOADED });
    const generatedMessage: boolean | MessageType<DefaultStreamChatGenerics> = message;
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        clearEditingState: clearEditingStateMock,
        clearQuotedMessageState: clearQuotedMessageStateMock,
        editing: generatedMessage,
        editMessage: editMessageMock,
        quotedMessage: false,
        sendMessage: sendMessageMock,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setImageUploads([images]);
      result.current.setMentionedUsers(['dummy1', 'dummy2']);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    expect(editMessageMock.mock.calls[0][0]).toMatchSnapshot();
    expect(clearEditingStateMock).toHaveBeenCalled();
    expect(result.current.sending.current).toBeFalsy();
    expect(result.current.fileUploads.length).toBe(0);
    expect(result.current.imageUploads.length).toBe(0);
    expect(result.current.mentionedUsers.length).toBe(0);
  });

  it('exit sendMessage when file upload status is uploaded successfully', () => {
    const files = [
      generateFileUploadPreview({ state: FileState.UPLOADED }),
      generateFileUploadPreview({ state: FileState.FINISHED, type: 'video/mp4' }),
      generateFileUploadPreview({ state: FileState.UPLOADED, type: 'audio/mp3' }),
      generateFileUploadPreview({ state: FileState.FINISHED, type: 'image/jpeg' }),
    ];
    const sendMessageMock = jest.fn();
    const clearQuotedMessageStateMock = jest.fn();
    const { result } = renderHook(() => useMessageInputContext(), {
      initialProps: {
        clearQuotedMessageState: clearQuotedMessageStateMock,
        editing: true,
        quotedMessage: false,
        sendMessage: sendMessageMock,
      },
      wrapper: Wrapper,
    });

    act(() => {
      result.current.setFileUploads(files);
      result.current.setMentionedUsers(['dummy1', 'dummy2']);
      result.current.setText('');
    });

    act(() => {
      result.current.sendMessage();
    });

    expect(sendMessageMock.mock.calls[0][0]).toMatchSnapshot();
    expect(clearQuotedMessageStateMock).toHaveBeenCalled();
    expect(result.current.sending.current).toBeFalsy();
    expect(result.current.fileUploads.length).toBe(0);
    expect(result.current.imageUploads.length).toBe(0);
    expect(result.current.mentionedUsers.length).toBe(0);
  });
});
