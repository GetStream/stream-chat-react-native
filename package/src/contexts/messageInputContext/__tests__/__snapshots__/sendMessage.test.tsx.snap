// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInputContext's sendMessage exit sendMessage when edit message is not boolean image upload status is uploaded successfully 1`] = `
Object {
  "attachments": Array [
    Object {
      "fallback": undefined,
      "image_url": undefined,
      "original_height": undefined,
      "original_width": undefined,
      "type": "image",
    },
  ],
  "created_at": "Sat Jul 02 2022 23:55:13 GMT+0530 (India Standard Time)",
  "html": "<p>regular</p>",
  "id": "7a85f744-cc89-4f82-a1d4-5456432cc8bf",
  "mentioned_users": Array [
    "dummy1",
    "dummy2",
  ],
  "quoted_message": undefined,
  "text": "",
  "type": "regular",
  "updated_at": "Sat Jul 02 2022 23:55:13 GMT+0530 (India Standard Time)",
  "user": Object {
    "banned": false,
    "created_at": "2020-04-27T13:39:49.331742Z",
    "id": "5d6f6322-567e-4e1e-af90-97ef1ed5cc23",
    "image": "fc86ddcb-bac4-400c-9afd-b0c0a1c0cd33",
    "name": "50cbdd0e-ca7e-4478-9e2c-be0f1ac6a995",
    "online": false,
    "role": "user",
    "updated_at": "2020-04-27T13:39:49.332087Z",
  },
}
`;

exports[`MessageInputContext's sendMessage exit sendMessage when file upload status is uploaded successfully 1`] = `
Object {
  "attachments": Array [
    Object {
      "asset_url": undefined,
      "file_size": undefined,
      "mime_type": "file",
      "title": "dummy.pdf",
      "type": "file",
    },
    Object {
      "asset_url": undefined,
      "duration": undefined,
      "file_size": undefined,
      "mime_type": "video/mp4",
      "thumb_url": undefined,
      "title": "dummy.pdf",
      "type": "video",
    },
    Object {
      "asset_url": undefined,
      "duration": undefined,
      "file_size": undefined,
      "mime_type": "audio/mp3",
      "title": "dummy.pdf",
      "type": "audio",
    },
    Object {
      "fallback": "dummy.pdf",
      "image_url": undefined,
      "type": "image",
    },
  ],
  "mentioned_users": Array [
    "dummy1",
    "dummy2",
  ],
  "parent_id": undefined,
  "quoted_message_id": undefined,
  "show_in_channel": undefined,
  "text": "",
}
`;

exports[`MessageInputContext's sendMessage exit sendMessage when image upload status is uploaded successfully 1`] = `
Object {
  "attachments": Array [
    Object {
      "fallback": undefined,
      "image_url": undefined,
      "original_height": undefined,
      "original_width": undefined,
      "type": "image",
    },
    Object {
      "fallback": undefined,
      "image_url": undefined,
      "original_height": undefined,
      "original_width": undefined,
      "type": "image",
    },
  ],
  "mentioned_users": Array [
    "dummy1",
    "dummy2",
  ],
  "parent_id": undefined,
  "quoted_message_id": undefined,
  "show_in_channel": undefined,
  "text": "",
}
`;
